hbase-client/src/main/java/org/apache/hadoop/hbase/client/AsyncTableImpl.java

```diff
@@ -311,7 +321,11 @@ public void onRegionError(RegionInfo region, Throwable error) {
 
       @Override
       public void onComplete() {
-        pool.execute(context.wrap(callback::onComplete));
+        pool.execute(context.wrap(() -> {
+          // Guarantee that onComplete() is called after all onRegionComplete()'s are called
+          regionCompletesInProgress.arriveAndAwaitAdvance();
```

Comment:
According to the Javadoc on this method, `onComplete` must implement a happens-after semantic for all `onRegionComplete` and all `onRegionError` invocations. This implies that there are multiple regions to complete AND multiple forms of completion, so I think that this can be solved with a phaser ; you need a more general purpose mutex.

Reply:
Good point that onRegionError should be included in the barrier. I'm guessing you mean to say "so I think that this _cannot_ be solved with a phaser," but I disagree. The phaser can be used in onRegionError just the same as onRegionComplete.

Reply:
Oh I see. Phaser is more powerful than I initially understood -- very cool! Yes, this looks better to me.